garden_ai/utils/_meta.py

```diff
@@ -27,5 +28,61 @@ def redef_in_main(obj):
     if obj.__module__ != "__main__":
         import __main__
 
-        s = inspect.getsource(obj)
-        exec(s, __main__.__dict__)
+        source = inspect.getsource(obj)
+        exec(source, __main__.__dict__)
+
+
+def exec_getsource(source, globals=None, locals=None):
+    """
+    helper: same as built in exec, but in a way that inspect.getsource can
+    still extract the source code of objects defined in the appropriate namespace.
+
+    This lets us use the default `DillCodeTextInspect` serialization
+    strategy with Globus Compute.
+    """
+    # https://stackoverflow.com/a/69668959
+    getlines = linecache.getlines
+
+    def monkey_patch(filename, module_globals=None):
+        if filename == "<string>":
+            return source.splitlines(keepends=True)
+        else:
+            return getlines(filename, module_globals)
+
+    # can now use inspect.getsource() on the result of exec() here
+    linecache.getlines = monkey_patch
```

Comment:
I think this leaves the monkey patch in place after this function has finished. (No cleanup `linecache.getlines = getlines` at the end.) Is that intentional?

Reply:
yeah, unfortunately. the monkeypatch needed to stay around long enough to still apply to globus compute's call to `getsource` in the serialization strategy

Reply:
I think morally it's ok here because it's only going to impact cases that would otherwise have been a straight up error from `getlines` but I don't love it